src/int.satyg

```diff
@@ -131,4 +147,44 @@ end = struct
         else
           panic (`Int.of-string: invalid string: `# ^ s))) 0 in
       sign * num
+
+  let to-string = arabic
+  let of-string-opt ?:radix-opt str =
+    let radix =
+      match radix-opt with
+      | None -> 10
+      | Some(i) -> i
+    in
+    let-rec make-i n m lst =
+      match lst with
+      | [] -> m
+      | x :: xs ->
+        let i =
+          n * x
+        in
+        (make-i (n * radix) (m + i) xs)
+    in
+    let cs = str |> String.to-list in
+    let to-i lst =
+      if List.exists Option.is-none lst then
+        None
+      else
+        let lst = List.show-opt lst in
+        if List.exists (fun i -> i >= radix) lst then
+          None
+        else
+          Some(make-i 1 0 lst)
+    in
+    match cs with
+    | [] -> None
+    | c::cs2 ->
+      if Eq.equal Char.eq c (Char.make `-`) then
+        let i-opt = cs2 |> List.reverse-map char-to-int |> to-i in
+        match i-opt with
+        | None -> None
+        | Some(i) -> Some(i * (-1))
+      else
+        cs |> List.reverse-map char-to-int |> to-i
```

Comment:
As opposed to the `of-string` function above, this implementation creates temporary lists during processing. I would like you to remove this by using List.fold-left just like in `of-string`:

```
      let num-opt = cs |> List.fold-left (fun acc-opt c ->
        acc-opt |> Option.map (fun acc ->
          (match char-to-int c with
          | None -> None
          | Some(i) ->
          if i < radix then
            Option.some (acc * radix + i)
         else
            Option.none)
        )) (Option.some 0) in
```

Reply:
Also, can you change the implementation of the `to-string` function above to like:

```
let to-string s = to-string-opt s |> Option.expect `nanka okasii`
```

Reply:
I fixed.